src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -947,87 +992,141 @@ static RegexNode CloneMultiWithoutFirstChar(RegexNode node)
+                            // the whole alternation, which means those failures should jump to
+                            // what "doneLabel" was defined as when starting the alternation.
+                            doneLabel = originalDoneLabel;
+                        }
 
                         // Emit the code for each branch.
                         EmitNode(node.Child(i));
 
+                        // Add this branch to the backtracking table.  At this point, either the child
+                        // had backtracking constructs, in which case doneLabel points to the last one
+                        // and that's where we'll want to jump to, or it doesn't, in which case doneLabel
+                        // still points to the nextBranch, which similarly is where we'll want to jump to.
+                        if (!isAtomic)
+                        {
+                            EmitRunstackResizeIfNeeded(2);
+                            writer.WriteLine($"{RunstackPush()} = {i};");
+                            writer.WriteLine($"{RunstackPush()} = {startingCrawlPos};");
+                            writer.WriteLine($"{RunstackPush()} = {startingRunTextPos};");
+                        }
+                        labelMap[i] = doneLabel;
+
                         // If we get here in the generated code, the branch completed successfully.
                         // Before jumping to the end, we need to zero out textSpanPos, so that no
                         // matter what the value is after the branch, whatever follows the alternate
                         // will see the same textSpanPos.
                         TransferTextSpanPosToRunTextPos();
-                        writer.WriteLine($"goto {doneAlternateLabel};");
+                        writer.WriteLine($"goto {matchLabel};");
+                        writer.WriteLine();
 
                         // Reset state for next branch and loop around to generate it.  This includes
                         // setting runtextpos back to what it was at the beginning of the alternation,
                         // updating textSpan to be the full length it was, and if there's a capture that
                         // needs to be reset, uncapturing it.
-                        MarkLabel(nextBranch, emitSemicolon: false);
-                        writer.WriteLine($"runtextpos = {startingRunTextPosName};");
-                        LoadTextSpanLocal(writer);
-                        textSpanPos = startingTextSpanPos;
-                        if (hasStartingCrawlpos)
+                        if (!isLastBranch)
                         {
-                            EmitUncaptureUntil(startingCrawlPos);
+                            MarkLabel(nextBranch!, emitSemicolon: false);
+                            writer.WriteLine($"runtextpos = {startingRunTextPos};");
+                            LoadTextSpanLocal(writer);
```

Comment:
I know this is not new, but do we actually need to reset the span here? I would assume you would hit this line only if an alternation didn't match before which AFAIK is the only thing that would have changed the span

Reply:
Even if it failed to match, it could have still advanced runtextpos by partially matching before failing.  Is that what you meant, or am I misunderstanding the question?

Reply:
Yes I think I see the scenario now. I just wasn't considering the backtrack case were this could be redefined later but then backtracked back to this alternation where we need to reset the span again.